plugins/catalog-backend/src/ingestion/processors/microsoftGraph/read.ts

```diff
@@ -0,0 +1,285 @@
+        user.id!,
+        // We are limiting the photo size, as users with full resolution photos
+        // can make the Backstage API slow
+        120,
+      );
+    };
+
+    picturePromises.push(loadPhoto());
+    entities.push(entity);
+  }
+
+  // Wait for all photos to be downloaded
+  await Promise.all(picturePromises);
+
+  return { users: entities };
+}
+
+export async function readMicrosoftGraphOrganizations(
+  client: MicrosoftGraphClient,
+): Promise<{
+  rootGroup?: GroupEntity; // With all relations empty
+}> {
+  const organizations = await client.getOrganization();
+
+  if (organizations) {
+    for (const entry of organizations) {
+      if (!entry.id || !entry.displayName) {
+        continue;
+      }
+
+      // For now we expect a single root orgranization
+      const name = normalizeEntityName(entry.displayName);
+      const rootGroup: GroupEntity = {
+        apiVersion: 'backstage.io/v1alpha1',
+        kind: 'Group',
+        metadata: {
+          name: name,
+          description: entry.displayName,
+          annotations: {
+            [MICROSOFT_GRAPH_TENANT_ID_ANNOTATION]: entry.id,
+          },
+        },
+        spec: {
+          type: 'root',
+          ancestors: [],
+          children: [],
+          descendants: [],
+        },
+      };
+
+      return { rootGroup };
+    }
+  }
+
+  return {};
+}
+
+export async function readMicrosoftGraphGroups(
+  client: MicrosoftGraphClient,
+  options?: { groupFilter?: string },
+): Promise<{
+  groups: GroupEntity[]; // With all relations empty
+  rootGroup: GroupEntity | undefined; // With all relations empty
+  groupMember: Map<string, Set<string>>;
+}> {
+  const groups: GroupEntity[] = [];
+  const groupMember: Map<string, Set<string>> = new Map();
+
+  const { rootGroup } = await readMicrosoftGraphOrganizations(client);
+
+  if (rootGroup) {
+    groupMember.set(rootGroup.metadata.name, new Set<string>());
+    groups.push(rootGroup);
+  }
+
+  const groupMemberPromises: Promise<void>[] = [];
+
+  for await (const group of client.getGroups({
+    filter: options?.groupFilter,
+    select: ['id', 'displayName', 'mailNickname'],
+  })) {
+    if (!group.id || !group.displayName) {
+      continue;
+    }
+
+    const name = normalizeEntityName(group.mailNickname || group.displayName);
+    const entity: GroupEntity = {
+      apiVersion: 'backstage.io/v1alpha1',
+      kind: 'Group',
+      metadata: {
+        name: name,
+        description: group.displayName,
+        annotations: {
+          [MICROSOFT_GRAPH_GROUP_ID_ANNOTATION]: group.id,
+        },
+      },
+      spec: {
+        type: 'team',
+        // TODO: We could include a group email and picture
+        ancestors: [],
+        children: [],
+        descendants: [],
+      },
+    };
+
+    const groupMembers = new Set<string>();
+
+    // Download the photos in parallel, otherwise it can take quite some time
```

Comment:
Members

Reply:
By the way, the catalog backend already has a dep on p-limit which is a handy thing to limit concurrency. Maybe you want to make all of these limited to like 10 at a time or something so you don't run into rate limits or so